sys/net/application_layer/gcoap/Kconfig

```diff
@@ -0,0 +1,144 @@
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+#
+menuconfig KCONFIG_MODULE_GCOAP
+    bool "Configure Gcoap"
+    depends on MODULE_GCOAP
+    help
+        Configure Gcoap module using Kconfig. If not set default values and
+        CFLAGS will be used.
+
+if KCONFIG_MODULE_GCOAP
+
+menu "Buffer Sizes"
+
+config GCOAP_PDU_BUF_SIZE
+    int "Request or response buffer size"
+    default 128
+    help
+        Size of the buffer used to build a CoAP request or response.
+
+config GCOAP_REQ_OPTIONS_BUF
+    int "Request options buffer size"
+    default 4
+    help
+        Reduce payload length by this value for a request.
+        Accommodates writing Content-Format option in gcoap_finish(). May be
+        set to zero if the function is not used.
+
+config GCOAP_RESP_OPTIONS_BUF
+    int "Response options buffer size"
+    default 4
+    help
+        Reduce payload length by this value for a response.
+        Accommodates writing Content-Format option in gcoap_finish(). May be
+        set to zero if the function is not used.
+
+config GCOAP_OBS_OPTIONS_BUF
+    int "Observe notification options buffer size"
+    default 4
+    help
+        Reduce payload length by this value for an observe notification.
+        Accommodates writing Content-Format option in gcoap_finish(). May be
+        set to zero if the function is not used.
+
+endmenu # Buffer Sizes
+
+menu "Observe options"
+
+config GCOAP_OBS_CLIENTS_MAX
+    int "Maximum number of Observe clients"
+    default 2
+
+config GCOAP_OBS_REGISTRATIONS_MAX
+    int "Maximum number of registrations for Observable resources"
+    default 2
+
+config GCOAP_OBS_VALUE_WIDTH
+    int "Width of the Observe option value for a notification"
+    default 3
+    range 1 3
+    help
+        This width, expressed in bytes, is used to determine the length of the
+        'tick' used to measure the time between observable changes to a
+        resource. A tick is expressed internally as GCOAP_OBS_TICK_EXPONENT,
+        which is the base 2 log value of the tick length in microseconds.
+
+        The canonical setting for the value is 3 (exponent 5), which results in
+        a tick length of 32 usec, per sec. 3.4, 4.4 of the RFC. Width 2
+        (exponent 16) results in a tick length of ~64 msec, and width 1
+        (exponent 24) results in a tick length of ~17 sec.
+
+        The tick length must be short enough so that the Observe value strictly
+        increases for each new notification. The purpose of the value is to
+        allow a client to detect message reordering withing the network latency
+        period (128 sec). For resources that change slowly, the reduced
+        message length is useful when packet size is limited.
+
+endmenu # Observe options
+
+menu "Timeouts and retries"
+
+config GCOAP_RECV_TIMEOUT
+    int "Incoming message timeout"
```

Comment:
general question on KConfig: are these converted to explicitly typed literals (`1000000LU`)?

Reply:
No, Kconfig does not add suffixes to these literals, the header file will just contain `1000000`. If needed, could we add a cast when using the macro?

Reply:
So all type checking by the compiler is removed?

Reply:
>So all type checking by the compiler is removed?

no, the compiler checks the type. What is removed is the possibility to set types for integer literals. This only affects CONFIG symbols that are used in arithmetic expressions. Also, the compiler still throws an error if integer widths do not match. Using an explicit cast (...) solves this problem.

Reply:
@kaspar030 ready to merge?

Reply:
Murdock shows green lights. Let's merge this now, but we can still continue the discussion in this thread for this particular kconfig property, if needed.